offsite-activity-signup/Code.gs

```diff
@@ -0,0 +1,346 @@
+ */
+function writeActivityRosters_(ss, activities) {
+  var sheetName = 'Activity rosters';
+  var sheet = ss.getSheetByName(sheetName);
+  if (!sheet) {
+    sheet = ss.insertSheet(sheetName);
+  }
+  sheet.clear();
+  var rows = [];
+  // Format each roster as a column in the sheet
+  activities.forEach(function(activity, colIndex) {
+    if (!rows[0]) {
+      rows[0] = fillArray_([], activities.length, '');
+    }
+    rows[0][colIndex] = activity.description;
+    activity.roster.forEach(function(attendee, rowIndex) {
+      if (!rows[rowIndex+1]) {
+        rows[rowIndex+1] = fillArray_([], activities.length, '');
+      }
+      rows[rowIndex+1][colIndex] = attendee.email;
+    });
+  });
+  rows.forEach(function(row) {
+    sheet.appendRow(row);
+  });
+  sheet.setFrozenRows(1);
+  sheet.getRange("1:1").setFontWeight("bold");
+  sheet.autoResizeColumns(1, sheet.getLastColumn());
+}
+
+/**
+ * Loads the activity schedule.
+ */
+function loadActivitySchedule_(ss) {
+  var timeZone = ss.getSpreadsheetTimeZone();
+  var sheet = ss.getSheetByName('Activity Schedule');
+  var rows = sheet.getSheetValues(2, 1, sheet.getLastRow() - 1, sheet.getLastRow());
+  var activities = rows.map(function(row, index) {
+    var name = row[0];
+    var startAt = new Date(row[1]);
+    var endAt = new Date(row[2]);
+    var capacity = parseInt(row[3]);
+    var formattedStartAt= Utilities.formatDate(startAt, timeZone, "EEE hh:mm a");
+    var formattedEndAt = Utilities.formatDate(endAt, timeZone, "hh:mm a");
+    var description = Utilities.formatString("%d: %s (%s-%s)", index, name, formattedStartAt, formattedEndAt);
+    return {
+      id: index,
+      name: name,
+      description: description,
+      capacity: capacity,
+      startAt: startAt,
+      endAt: endAt,
+      roster: []
+    }
+  });
+  return activities;
+}
+
+/**
+ * Loads the attendeee response data
+ */
+function loadAttendeeResponses_(ss, allActivityIds) {
+  var sheet = findResponseSheetForForm_(ss);
+  if (!sheet || sheet.getLastRow() == 1) {
+    return undefined;
+  }  
+  var rows = sheet.getSheetValues(2, 1, sheet.getLastRow() - 1, sheet.getLastRow());
+  var attendees = rows.map(function(row) {
+    var _ = row.shift(); // Ignore timestamp
+    var email = row.shift();
+    var autoAssign = row.pop(); 
+    var preferences = row.map(function(pick) {
+      var match = pick.match(/(\d+):.*/);
+      if (!match) {
+        return undefined;
+      }
+      return parseInt(match[1]);
+    });    
+    if (autoAssign == 'Yes') {
+      // If auto assigning additional activites, append a randomized
+      // list of all the activities. These will then be considered
+      // as if the attendee ranked them.
+      var additionalChoices = shuffleArray_(allActivityIds.slice(0));
+      preferences = preferences.concat(additionalChoices);
+    }    
+    return {
+      email: email,
+      preferences: preferences,
+      assigned: []
+    };
+  });
+  return attendees;
+}
+
+/**
```

Comment:
I'm not sure if the solutions should also include helpers to generate dummy data. I see a case for an against. Perhaps worth discussing in a meeting / chat.

Reply:
We discussed in the meeting, and the decision is to include code like this, to increase demoability. But we thought it might be best to move into a separate file though, to make it clear what code is ancillary setup.